File: code/dotnet/FactSetConcordance/v2/src/FactSet.SDK.FactSetConcordance/Model/Universe.cs
/*
 * FactSet Concordance API
 *
 * The FactSet Concordance API provides the ability to programmatically look up several types of entities by their name and other attributes. It attempts to match attributes that you supply for the entity and returns as many match candidates as it can find. The result is a \"concorded\" FactSet Entity Permanent identifier.  The Entity Identifier can then be used as input throughout FactSet's Content soultions, such as Content APIs and Standard DataFeeds.<p> There are two types of workflows supported in the API. The first workflow type follows traditional API conventions in that a request is accepted and a response is returned synchronously. The second workflow type involves a background task. With this workflow, a task is created for uploading a set of queries, and then the client polls for the task to complete. Once the task has completed, the client retrieves the concordance results throught the decisions endpoint.</p> 
 *
 * The version of the OpenAPI document: 2.0.0
 * Contact: api@factset.com
 * Generated by: https://github.com/openapitools/openapi-generator.git
 */


using System;
using System.Collections;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;
using System.IO;
using System.Runtime.Serialization;
using System.Text;
using System.Text.RegularExpressions;
using Newtonsoft.Json;
using Newtonsoft.Json.Converters;
using Newtonsoft.Json.Linq;
using System.ComponentModel.DataAnnotations;
using OpenAPIDateConverter = FactSet.SDK.FactSetConcordance.Client.OpenAPIDateConverter;

namespace FactSet.SDK.FactSetConcordance.Model
{
    /// <summary>
    /// Metadata about a single universe
    /// </summary>
    [DataContract(Name = "Universe")]
    public partial class Universe : IEquatable<Universe>, IValidatableObject
    {
        /// <summary>
        /// Initializes a new instance of the <see cref="Universe" /> class.
        /// </summary>
        /// <param name="universeId">The universe identifier.</param>
        /// <param name="universeName">The user specified universe name.</param>
        /// <param name="universeDescription">The user specified universe description.</param>
        /// <param name="universeType">The user specified universe entity..</param>
        /// <param name="selfManaged">False if the universe is managed by FactSet Managed Service, True otherwise.</param>
        /// <param name="createdBy">The user who created the universe.</param>
        /// <param name="createdOn">When the universe was created.</param>
        /// <param name="updatedBy">The user who most recently updated the universe metadata.</param>
        /// <param name="updatedOn">When the universe metadata was most recently updated.</param>
        public Universe(int? universeId = default(int?), string universeName = default(string), string universeDescription = default(string), string universeType = default(string), bool? selfManaged = default(bool?), string createdBy = default(string), DateTime? createdOn = default(DateTime?), string updatedBy = default(string), DateTime? updatedOn = default(DateTime?))
        {
            this.UniverseId = universeId;
            this.UniverseName = universeName;
            this.UniverseDescription = universeDescription;
            this.UniverseType = universeType;
            this.SelfManaged = selfManaged;
            this.CreatedBy = createdBy;
            this.CreatedOn = createdOn;
            this.UpdatedBy = updatedBy;
            this.UpdatedOn = updatedOn;
        }

        /// <summary>
        /// The universe identifier
        /// </summary>
        /// <value>The universe identifier</value>
        [DataMember(Name = "universeId", EmitDefaultValue = true)]
        public int? UniverseId { get; set; }

        /// <summary>
        /// The user specified universe name
        /// </summary>
        /// <value>The user specified universe name</value>
        [DataMember(Name = "universeName", EmitDefaultValue = true)]
        public string UniverseName { get; set; }

        /// <summary>
        /// The user specified universe description
        /// </summary>
        /// <value>The user specified universe description</value>
        [DataMember(Name = "universeDescription", EmitDefaultValue = true)]
        public string UniverseDescription { get; set; }

        /// <summary>
        /// The user specified universe entity.
        /// </summary>
        /// <value>The user specified universe entity.</value>
        [DataMember(Name = "universeType", EmitDefaultValue = true)]
        public string UniverseType { get; set; }

        /// <summary>
        /// False if the universe is managed by FactSet Managed Service, True otherwise
        /// </summary>
        /// <value>False if the universe is managed by FactSet Managed Service, True otherwise</value>
        [DataMember(Name = "selfManaged", EmitDefaultValue = true)]
        public bool? SelfManaged { get; set; }

        /// <summary>
        /// The user who created the universe
        /// </summary>
        /// <value>The user who created the universe</value>
        [DataMember(Name = "createdBy", EmitDefaultValue = true)]
        public string CreatedBy { get; set; }

        /// <summary>
        /// When the universe was created
        /// </summary>
        /// <value>When the universe was created</value>
        [DataMember(Name = "createdOn", EmitDefaultValue = true)]
        public DateTime? CreatedOn { get; set; }

        /// <summary>
        /// The user who most recently updated the universe metadata
        /// </summary>
        /// <value>The user who most recently updated the universe metadata</value>
        [DataMember(Name = "updatedBy", EmitDefaultValue = true)]
        public string UpdatedBy { get; set; }

        /// <summary>
        /// When the universe metadata was most recently updated
        /// </summary>
        /// <value>When the universe metadata was most recently updated</value>
        [DataMember(Name = "updatedOn", EmitDefaultValue = true)]
        public DateTime? UpdatedOn { get; set; }

        /// <summary>
        /// Returns the string presentation of the object
        /// </summary>
        /// <returns>String presentation of the object</returns>
        public override string ToString()
        {
            StringBuilder sb = new StringBuilder();
            sb.Append("class Universe {\n");
            sb.Append("  UniverseId: ").Append(UniverseId).Append("\n");
            sb.Append("  UniverseName: ").Append(UniverseName).Append("\n");
            sb.Append("  UniverseDescription: ").Append(UniverseDescription).Append("\n");
            sb.Append("  UniverseType: ").Append(UniverseType).Append("\n");
            sb.Append("  SelfManaged: ").Append(SelfManaged).Append("\n");
            sb.Append("  CreatedBy: ").Append(CreatedBy).Append("\n");
            sb.Append("  CreatedOn: ").Append(CreatedOn).Append("\n");
            sb.Append("  UpdatedBy: ").Append(UpdatedBy).Append("\n");
            sb.Append("  UpdatedOn: ").Append(UpdatedOn).Append("\n");
            sb.Append("}\n");
            return sb.ToString();
        }

        /// <summary>
        /// Returns the JSON string presentation of the object
        /// </summary>
        /// <returns>JSON string presentation of the object</returns>
        public virtual string ToJson()
        {
            return Newtonsoft.Json.JsonConvert.SerializeObject(this, Newtonsoft.Json.Formatting.Indented);
        }

        /// <summary>
        /// Returns true if objects are equal
        /// </summary>
        /// <param name="input">Object to be compared</param>
        /// <returns>Boolean</returns>
        public override bool Equals(object input)
        {
            return this.Equals(input as Universe);
        }

        /// <summary>
        /// Returns true if Universe instances are equal
        /// </summary>
        /// <param name="input">Instance of Universe to be compared</param>
        /// <returns>Boolean</returns>
        public bool Equals(Universe input)
        {
            if (input == null)
            {
                return false;
            }
            return 
                (
                    this.UniverseId == input.UniverseId ||
                    (this.UniverseId != null &&
                    this.UniverseId.Equals(input.UniverseId))
                ) && 
                (
                    this.UniverseName == input.UniverseName ||
                    (this.UniverseName != null &&
                    this.UniverseName.Equals(input.UniverseName))
                ) && 
                (
                    this.UniverseDescription == input.UniverseDescription ||
                    (this.UniverseDescription != null &&
                    this.UniverseDescription.Equals(input.UniverseDescription))
                ) && 
                (
                    this.UniverseType == input.UniverseType ||
                    (this.UniverseType != null &&
                    this.UniverseType.Equals(input.UniverseType))
                ) && 
                (
                    this.SelfManaged == input.SelfManaged ||
                    (this.SelfManaged != null &&
                    this.SelfManaged.Equals(input.SelfManaged))
                ) && 
                (
                    this.CreatedBy == input.CreatedBy ||
                    (this.CreatedBy != null &&
                    this.CreatedBy.Equals(input.CreatedBy))
                ) && 
                (
                    this.CreatedOn == input.CreatedOn ||
                    (this.CreatedOn != null &&
                    this.CreatedOn.Equals(input.CreatedOn))
                ) && 
                (
                    this.UpdatedBy == input.UpdatedBy ||
                    (this.UpdatedBy != null &&
                    this.UpdatedBy.Equals(input.UpdatedBy))
                ) && 
                (
                    this.UpdatedOn == input.UpdatedOn ||
                    (this.UpdatedOn != null &&
                    this.UpdatedOn.Equals(input.UpdatedOn))
                );
        }

        /// <summary>
        /// Gets the hash code
        /// </summary>
        /// <returns>Hash code</returns>
        public override int GetHashCode()
        {
            unchecked // Overflow is fine, just wrap
            {
                int hashCode = 41;
                if (this.UniverseId != null)
                {
                    hashCode = (hashCode * 59) + this.UniverseId.GetHashCode();
                }
                if (this.UniverseName != null)
                {
                    hashCode = (hashCode * 59) + this.UniverseName.GetHashCode();
                }
                if (this.UniverseDescription != null)
                {
                    hashCode = (hashCode * 59) + this.UniverseDescription.GetHashCode();
                }
                if (this.UniverseType != null)
                {
                    hashCode = (hashCode * 59) + this.UniverseType.GetHashCode();
                }
                if (this.SelfManaged != null)
                {
                    hashCode = (hashCode * 59) + this.SelfManaged.GetHashCode();
                }
                if (this.CreatedBy != null)
                {
                    hashCode = (hashCode * 59) + this.CreatedBy.GetHashCode();
                }
                if (this.CreatedOn != null)
                {
                    hashCode = (hashCode * 59) + this.CreatedOn.GetHashCode();
                }
                if (this.UpdatedBy != null)
                {
                    hashCode = (hashCode * 59) + this.UpdatedBy.GetHashCode();
                }
                if (this.UpdatedOn != null)
                {
                    hashCode = (hashCode * 59) + this.UpdatedOn.GetHashCode();
                }
                return hashCode;
            }
        }

        /// <summary>
        /// To validate all properties of the instance
        /// </summary>
        /// <param name="validationContext">Validation context</param>
        /// <returns>Validation Result</returns>
        public IEnumerable<System.ComponentModel.DataAnnotations.ValidationResult> Validate(ValidationContext validationContext)
        {
            yield break;
        }
    }

}

File: code/java/FactSetConcordance/v2/src/main/java/com/factset/sdk/FactSetConcordance/models/Universe.java
/*
 * FactSet Concordance API
 *   The FactSet Concordance API enables Application Developers and Data Scientists to programmatically discover the FactSet Identifier for a specific Entity or Person based on attributes, such as name, URL, and location.FactSet Identifiers are mapped to industry-standard identifiers, as well as to a comprehensive set of reference data and unique content, to enhance the ability to quickly connect content sets.<p>     Using the submitted attributes, the FactSet Concordance API leverages FactSet’s extensive Entity Master, People, and Symbology databases to return potential candidate matches and a proposed match. The result is a concorded FactSet Identifier that can then be used as input throughout FactSet's expanding catalog of Content using our Content APIs or Standard DataFeeds.</p>    <b>There are two types of workflows supported in the API:</b>    <b>Entity & People Match:</b> The first workflow follows traditional API conventions in that a request is accepted and a response is returned synchronously enabling the concordance of up to 25 names in a single request. The response returns a list of 20 candidates as well as a proposed match.    <b>Entity & People Match - Bulk:</b> The second workflow allows you to input a large list of names within a .CSV file in a single request With this workflow, a task is created for uploading a set of queries. Once the task has completed, the client retrieves the concordance results through the /entity-decisions or /people-decisions endpoint. Note that in the Bulk workflow, only the matches for the records are returned. Candidates are not included.    Once the service has provided a list of concorded FactSet Identifiers, you can then use our Entity Mappings and People Mappings endpoints to further review the universe of mapped identifiers or modify existing records. </p> 
 *
 * The version of the OpenAPI document: 2.7.0
 * Contact: api@factset.com
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */


package com.factset.sdk.FactSetConcordance.models;

import java.util.Objects;
import java.util.Arrays;
import java.util.Map;
import java.util.HashMap;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonTypeName;
import com.fasterxml.jackson.annotation.JsonValue;
import io.swagger.annotations.ApiModel;
import io.swagger.annotations.ApiModelProperty;
import java.time.OffsetDateTime;
import org.openapitools.jackson.nullable.JsonNullable;
import com.fasterxml.jackson.annotation.JsonIgnore;
import org.openapitools.jackson.nullable.JsonNullable;
import java.util.NoSuchElementException;
import java.io.Serializable;
import com.fasterxml.jackson.annotation.JsonPropertyOrder;
import com.factset.sdk.FactSetConcordance.JSON;


/**
 * Metadata about a single universe
 */
@ApiModel(description = "Metadata about a single universe")
@JsonPropertyOrder({
  Universe.JSON_PROPERTY_UNIVERSE_ID,
  Universe.JSON_PROPERTY_UNIVERSE_NAME,
  Universe.JSON_PROPERTY_UNIVERSE_DESCRIPTION,
  Universe.JSON_PROPERTY_UNIVERSE_TYPE,
  Universe.JSON_PROPERTY_SELF_MANAGED,
  Universe.JSON_PROPERTY_CREATED_BY,
  Universe.JSON_PROPERTY_CREATED_ON,
  Universe.JSON_PROPERTY_UPDATED_BY,
  Universe.JSON_PROPERTY_UPDATED_ON
})
@javax.annotation.Generated(value = "org.openapitools.codegen.languages.JavaClientCodegen")
public class Universe implements Serializable {
  private static final long serialVersionUID = 1L;

  public static final String JSON_PROPERTY_UNIVERSE_ID = "universeId";
  private JsonNullable<Integer> universeId = JsonNullable.<Integer>undefined();

  public static final String JSON_PROPERTY_UNIVERSE_NAME = "universeName";
  private JsonNullable<String> universeName = JsonNullable.<String>undefined();

  public static final String JSON_PROPERTY_UNIVERSE_DESCRIPTION = "universeDescription";
  private JsonNullable<String> universeDescription = JsonNullable.<String>undefined();

  public static final String JSON_PROPERTY_UNIVERSE_TYPE = "universeType";
  private JsonNullable<String> universeType = JsonNullable.<String>undefined();

  public static final String JSON_PROPERTY_SELF_MANAGED = "selfManaged";
  private JsonNullable<Boolean> selfManaged = JsonNullable.<Boolean>undefined();

  public static final String JSON_PROPERTY_CREATED_BY = "createdBy";
  private JsonNullable<String> createdBy = JsonNullable.<String>undefined();

  public static final String JSON_PROPERTY_CREATED_ON = "createdOn";
  private JsonNullable<OffsetDateTime> createdOn = JsonNullable.<OffsetDateTime>undefined();

  public static final String JSON_PROPERTY_UPDATED_BY = "updatedBy";
  private JsonNullable<String> updatedBy = JsonNullable.<String>undefined();

  public static final String JSON_PROPERTY_UPDATED_ON = "updatedOn";
  private JsonNullable<OffsetDateTime> updatedOn = JsonNullable.<OffsetDateTime>undefined();

  public Universe() { 
  }

  public Universe universeId(Integer universeId) {
    this.universeId = JsonNullable.<Integer>of(universeId);
    return this;
  }

   /**
   * The universe identifier
   * @return universeId
  **/
  @javax.annotation.Nullable
  @ApiModelProperty(example = "1", value = "The universe identifier")
  @JsonIgnore

  public Integer getUniverseId() {
        return universeId.orElse(null);
  }

  @JsonProperty(JSON_PROPERTY_UNIVERSE_ID)
  @JsonInclude(value = JsonInclude.Include.USE_DEFAULTS)

  public JsonNullable<Integer> getUniverseId_JsonNullable() {
    return universeId;
  }
  
  @JsonProperty(JSON_PROPERTY_UNIVERSE_ID)
  public void setUniverseId_JsonNullable(JsonNullable<Integer> universeId) {
    this.universeId = universeId;
  }

  public void setUniverseId(Integer universeId) {
    this.universeId = JsonNullable.<Integer>of(universeId);
  }


  public Universe universeName(String universeName) {
    this.universeName = JsonNullable.<String>of(universeName);
    return this;
  }

   /**
   * The user specified universe name
   * @return universeName
  **/
  @javax.annotation.Nullable
  @ApiModelProperty(example = "My Universe", value = "The user specified universe name")
  @JsonIgnore

  public String getUniverseName() {
        return universeName.orElse(null);
  }

  @JsonProperty(JSON_PROPERTY_UNIVERSE_NAME)
  @JsonInclude(value = JsonInclude.Include.USE_DEFAULTS)

  public JsonNullable<String> getUniverseName_JsonNullable() {
    return universeName;
  }
  
  @JsonProperty(JSON_PROPERTY_UNIVERSE_NAME)
  public void setUniverseName_JsonNullable(JsonNullable<String> universeName) {
    this.universeName = universeName;
  }

  public void setUniverseName(String universeName) {
    this.universeName = JsonNullable.<String>of(universeName);
  }


  public Universe universeDescription(String universeDescription) {
    this.universeDescription = JsonNullable.<String>of(universeDescription);
    return this;
  }

   /**
   * The user specified universe description
   * @return universeDescription
  **/
  @javax.annotation.Nullable
  @ApiModelProperty(example = "This is a universe", value = "The user specified universe description")
  @JsonIgnore

  public String getUniverseDescription() {
        return universeDescription.orElse(null);
  }

  @JsonProperty(JSON_PROPERTY_UNIVERSE_DESCRIPTION)
  @JsonInclude(value = JsonInclude.Include.USE_DEFAULTS)

  public JsonNullable<String> getUniverseDescription_JsonNullable() {
    return universeDescription;
  }
  
  @JsonProperty(JSON_PROPERTY_UNIVERSE_DESCRIPTION)
  public void setUniverseDescription_JsonNullable(JsonNullable<String> universeDescription) {
    this.universeDescription = universeDescription;
  }

  public void setUniverseDescription(String universeDescription) {
    this.universeDescription = JsonNullable.<String>of(universeDescription);
  }


  public Universe universeType(String universeType) {
    this.universeType = JsonNullable.<String>of(universeType);
    return this;
  }

   /**
   * The user specified universe entity.
   * @return universeType
  **/
  @javax.annotation.Nullable
  @ApiModelProperty(example = "ENTITY", value = "The user specified universe entity.")
  @JsonIgnore

  public String getUniverseType() {
        return universeType.orElse(null);
  }

  @JsonProperty(JSON_PROPERTY_UNIVERSE_TYPE)
  @JsonInclude(value = JsonInclude.Include.USE_DEFAULTS)

  public JsonNullable<String> getUniverseType_JsonNullable() {
    return universeType;
  }
  
  @JsonProperty(JSON_PROPERTY_UNIVERSE_TYPE)
  public void setUniverseType_JsonNullable(JsonNullable<String> universeType) {
    this.universeType = universeType;
  }

  public void setUniverseType(String universeType) {
    this.universeType = JsonNullable.<String>of(universeType);
  }


  public Universe selfManaged(Boolean selfManaged) {
    this.selfManaged = JsonNullable.<Boolean>of(selfManaged);
    return this;
  }

   /**
   * False if the universe is managed by FactSet Managed Service, True otherwise
   * @return selfManaged
  **/
  @javax.annotation.Nullable
  @ApiModelProperty(example = "true", value = "False if the universe is managed by FactSet Managed Service, True otherwise")
  @JsonIgnore

  public Boolean getSelfManaged() {
        return selfManaged.orElse(null);
  }

  @JsonProperty(JSON_PROPERTY_SELF_MANAGED)
  @JsonInclude(value = JsonInclude.Include.USE_DEFAULTS)

  public JsonNullable<Boolean> getSelfManaged_JsonNullable() {
    return selfManaged;
  }
  
  @JsonProperty(JSON_PROPERTY_SELF_MANAGED)
  public void setSelfManaged_JsonNullable(JsonNullable<Boolean> selfManaged) {
    this.selfManaged = selfManaged;
  }

  public void setSelfManaged(Boolean selfManaged) {
    this.selfManaged = JsonNullable.<Boolean>of(selfManaged);
  }


  public Universe createdBy(String createdBy) {
    this.createdBy = JsonNullable.<String>of(createdBy);
    return this;
  }

   /**
   * The user who created the universe
   * @return createdBy
  **/
  @javax.annotation.Nullable
  @ApiModelProperty(example = "FDSQAR_C-000000", value = "The user who created the universe")
  @JsonIgnore

  public String getCreatedBy() {
        return createdBy.orElse(null);
  }

  @JsonProperty(JSON_PROPERTY_CREATED_BY)
  @JsonInclude(value = JsonInclude.Include.USE_DEFAULTS)

  public JsonNullable<String> getCreatedBy_JsonNullable() {
    return createdBy;
  }
  
  @JsonProperty(JSON_PROPERTY_CREATED_BY)
  public void setCreatedBy_JsonNullable(JsonNullable<String> createdBy) {
    this.createdBy = createdBy;
  }

  public void setCreatedBy(String createdBy) {
    this.createdBy = JsonNullable.<String>of(createdBy);
  }


  public Universe createdOn(OffsetDateTime createdOn) {
    this.createdOn = JsonNullable.<OffsetDateTime>of(createdOn);
    return this;
  }

   /**
   * When the universe was created
   * @return createdOn
  **/
  @javax.annotation.Nullable
  @ApiModelProperty(value = "When the universe was created")
  @JsonIgnore

  public OffsetDateTime getCreatedOn() {
        return createdOn.orElse(null);
  }

  @JsonProperty(JSON_PROPERTY_CREATED_ON)
  @JsonInclude(value = JsonInclude.Include.USE_DEFAULTS)

  public JsonNullable<OffsetDateTime> getCreatedOn_JsonNullable() {
    return createdOn;
  }
  
  @JsonProperty(JSON_PROPERTY_CREATED_ON)
  public void setCreatedOn_JsonNullable(JsonNullable<OffsetDateTime> createdOn) {
    this.createdOn = createdOn;
  }

  public void setCreatedOn(OffsetDateTime createdOn) {
    this.createdOn = JsonNullable.<OffsetDateTime>of(createdOn);
  }


  public Universe updatedBy(String updatedBy) {
    this.updatedBy = JsonNullable.<String>of(updatedBy);
    return this;
  }

   /**
   * The user who most recently updated the universe metadata
   * @return updatedBy
  **/
  @javax.annotation.Nullable
  @ApiModelProperty(example = "FDSQAR_C-000000", value = "The user who most recently updated the universe metadata")
  @JsonIgnore

  public String getUpdatedBy() {
        return updatedBy.orElse(null);
  }

  @JsonProperty(JSON_PROPERTY_UPDATED_BY)
  @JsonInclude(value = JsonInclude.Include.USE_DEFAULTS)

  public JsonNullable<String> getUpdatedBy_JsonNullable() {
    return updatedBy;
  }
  
  @JsonProperty(JSON_PROPERTY_UPDATED_BY)
  public void setUpdatedBy_JsonNullable(JsonNullable<String> updatedBy) {
    this.updatedBy = updatedBy;
  }

  public void setUpdatedBy(String updatedBy) {
    this.updatedBy = JsonNullable.<String>of(updatedBy);
  }


  public Universe updatedOn(OffsetDateTime updatedOn) {
    this.updatedOn = JsonNullable.<OffsetDateTime>of(updatedOn);
    return this;
  }

   /**
   * When the universe metadata was most recently updated
   * @return updatedOn
  **/
  @javax.annotation.Nullable
  @ApiModelProperty(value = "When the universe metadata was most recently updated")
  @JsonIgnore

  public OffsetDateTime getUpdatedOn() {
        return updatedOn.orElse(null);
  }

  @JsonProperty(JSON_PROPERTY_UPDATED_ON)
  @JsonInclude(value = JsonInclude.Include.USE_DEFAULTS)

  public JsonNullable<OffsetDateTime> getUpdatedOn_JsonNullable() {
    return updatedOn;
  }
  
  @JsonProperty(JSON_PROPERTY_UPDATED_ON)
  public void setUpdatedOn_JsonNullable(JsonNullable<OffsetDateTime> updatedOn) {
    this.updatedOn = updatedOn;
  }

  public void setUpdatedOn(OffsetDateTime updatedOn) {
    this.updatedOn = JsonNullable.<OffsetDateTime>of(updatedOn);
  }


  /**
   * Return true if this Universe object is equal to o.
   */
  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    Universe universe = (Universe) o;
    return equalsNullable(this.universeId, universe.universeId) &&
        equalsNullable(this.universeName, universe.universeName) &&
        equalsNullable(this.universeDescription, universe.universeDescription) &&
        equalsNullable(this.universeType, universe.universeType) &&
        equalsNullable(this.selfManaged, universe.selfManaged) &&
        equalsNullable(this.createdBy, universe.createdBy) &&
        equalsNullable(this.createdOn, universe.createdOn) &&
        equalsNullable(this.updatedBy, universe.updatedBy) &&
        equalsNullable(this.updatedOn, universe.updatedOn);
  }

  private static <T> boolean equalsNullable(JsonNullable<T> a, JsonNullable<T> b) {
    return a == b || (a != null && b != null && a.isPresent() && b.isPresent() && Objects.deepEquals(a.get(), b.get()));
  }

  @Override
  public int hashCode() {
    return Objects.hash(hashCodeNullable(universeId), hashCodeNullable(universeName), hashCodeNullable(universeDescription), hashCodeNullable(universeType), hashCodeNullable(selfManaged), hashCodeNullable(createdBy), hashCodeNullable(createdOn), hashCodeNullable(updatedBy), hashCodeNullable(updatedOn));
  }

  private static <T> int hashCodeNullable(JsonNullable<T> a) {
    if (a == null) {
      return 1;
    }
    return a.isPresent() ? Arrays.deepHashCode(new Object[]{a.get()}) : 31;
  }

  @Override
  public String toString() {
    StringBuilder sb = new StringBuilder();
    sb.append("class Universe {\n");
    sb.append("    universeId: ").append(toIndentedString(universeId)).append("\n");
    sb.append("    universeName: ").append(toIndentedString(universeName)).append("\n");
    sb.append("    universeDescription: ").append(toIndentedString(universeDescription)).append("\n");
    sb.append("    universeType: ").append(toIndentedString(universeType)).append("\n");
    sb.append("    selfManaged: ").append(toIndentedString(selfManaged)).append("\n");
    sb.append("    createdBy: ").append(toIndentedString(createdBy)).append("\n");
    sb.append("    createdOn: ").append(toIndentedString(createdOn)).append("\n");
    sb.append("    updatedBy: ").append(toIndentedString(updatedBy)).append("\n");
    sb.append("    updatedOn: ").append(toIndentedString(updatedOn)).append("\n");
    sb.append("}");
    return sb.toString();
  }

  /**
   * Convert the given object to string with each line indented by 4 spaces
   * (except the first line).
   */
  private String toIndentedString(Object o) {
    if (o == null) {
      return "null";
    }
    return o.toString().replace("\n", "\n    ");
  }

}

